go/storage/mkvs/db/pathbadger/pathbadger.go

```diff
@@ -0,0 +1,943 @@
+		err = item.Value(func(data []byte) error {
+			return cbor.UnmarshalTrusted(data, &d.meta.value)
+		})
+		if err != nil {
+			return err
+		}
+
+		if d.meta.value.Version != dbVersion {
+			return fmt.Errorf("incompatible database version (expected: %d got: %d)",
+				dbVersion,
+				d.meta.value.Version,
+			)
+		}
+		if !d.meta.value.Namespace.Equal(&d.namespace) {
+			return fmt.Errorf("incompatible namespace (expected: %s got: %s)",
+				d.namespace,
+				d.meta.value.Namespace,
+			)
+		}
+		return nil
+	case badger.ErrKeyNotFound:
+	default:
+		return err
+	}
+
+	// No metadata exists, create some.
+	d.meta.value.Version = dbVersion
+	d.meta.value.Namespace = d.namespace
+	if err = d.meta.save(tx); err != nil {
+		return err
+	}
+
+	return tx.CommitAt(tsMetadata, nil)
+}
+
+func (d *badgerNodeDB) sanityCheckNamespace(ns *common.Namespace) error {
+	if !ns.Equal(&d.namespace) {
+		return api.ErrBadNamespace
+	}
+	return nil
+}
+
+func (d *badgerNodeDB) checkRootExists(tx *badger.Txn, root node.Root) error {
+	rootHash := api.TypedHashFromRoot(root)
+	if _, err := tx.Get(rootNodeKeyFmt.Encode(root.Version, &rootHash)); err != nil {
+		switch err {
+		case badger.ErrKeyNotFound:
+			return api.ErrRootNotFound
+		default:
+			d.logger.Error("failed to check root existence",
+				"err", err,
+			)
+			return fmt.Errorf("mkvs/pathbadger: failed to check root existence while getting node from backing store: %w", err)
+		}
+	}
+	return nil
+}
+
+// Implements api.NodeDB.
+func (d *badgerNodeDB) GetLatestVersion() (uint64, bool) {
+	return d.meta.getLastFinalizedVersion()
+}
+
+// Implements api.NodeDB.
+func (d *badgerNodeDB) GetEarliestVersion() uint64 {
+	return d.meta.getEarliestVersion()
+}
+
+// Implements api.NodeDB.
+func (d *badgerNodeDB) GetRootsForVersion(version uint64) (roots []node.Root, err error) {
+	// If the version is earlier than the earliest version, we don't have the roots.
+	if version < d.meta.getEarliestVersion() {
+		return nil, nil
+	}
+
+	tx := d.db.NewTransactionAt(versionToTs(version), false)
+	defer tx.Discard()
+
+	prefix := rootNodeKeyFmt.Encode(version)
+	it := tx.NewIterator(badger.IteratorOptions{Prefix: prefix})
+	defer it.Close()
+
+	for it.Rewind(); it.Valid(); it.Next() {
+		var (
+			v        uint64
+			rootHash api.TypedHash
+		)
+		if !rootNodeKeyFmt.Decode(it.Item().Key(), &v, &rootHash) {
+			panic("mkvs/pathbadger: corrupted key")
+		}
+
+		roots = append(roots, node.Root{
+			Namespace: d.namespace,
+			Version:   version,
+			Type:      rootHash.Type(),
+			Hash:      rootHash.Hash(),
+		})
+	}
+	return
+}
+
+// Implements api.NodeDB.
+func (d *badgerNodeDB) HasRoot(root node.Root) bool {
+	if err := d.sanityCheckNamespace(&root.Namespace); err != nil {
+		return false
+	}
+
+	// An empty root is always implicitly present.
+	if root.Hash.IsEmpty() {
+		return true
+	}
+
+	// If the version is earlier than the earliest version, we don't have the root.
+	if root.Version < d.meta.getEarliestVersion() {
```

Comment:
```suggestion
	if _, ok := d.meta.getLastFinalizedVersion(); !ok {
		return false
	}
	if root.Version < d.meta.getEarliestVersion() {
```

Reply:
Why? It should be possible to read non-finalized roots. Otherwise this will cause the node to re-download roots that it already has locally.